## Gather

In [186]:
import pandas as pd

In [84]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

### General Input Data Information

Check samples from all dataframes and view the generic info for each.

In [19]:
patients.head(3)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2       402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8

In [20]:
treatments.head(3)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN

In [21]:
adverse_reactions.head(3)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia

In [22]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [23]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


adverse_reactions.info()

### Duplicates and missing values

Check which columns are duplicated over the three dataframes. 

In [25]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

Get the columns of the patients dataframe and check for missing values. 

In [26]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [27]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [28]:
#Check if we have duplicate patients (based on surname)
patients.surname.value_counts()

Doe              6
Jakobsen         3
Taylor           3
Gersten          2
Woźniak          2
Parker           2
Lund             2
Cindrić          2
Kadyrov          2
Berg             2
Grímsdóttir      2
Lương            2
Correia          2
Ogochukwu        2
Dratchev         2
Tucker           2
Lâm              2
Souza            2
Schiavone        2
Silva            2
Cabrera          2
Bùi              2
Hueber           2
Collins          2
Tạ               2
Kowalczyk        2
Liễu             2
Aranda           2
Batukayev        2
Nilsen           2
                ..
Koldenhof        1
Bouw             1
Debord           1
Mironov          1
Mayberry         1
Lorenzo          1
She              1
Holden           1
Zimmerman        1
Poulsen          1
Labrosse         1
Bogolyubova      1
Carreiro         1
Dandonneau       1
Enríquez         1
Baumgaertner     1
Czerwinska       1
Pérez            1
Grunewald        1
Vaara            1
Nordin           1
Tromp       

In [30]:
#Check for duplicates in the patient addresses
patients.address.value_counts()

123 Main Street               6
2476 Fulton Street            2
648 Old Dear Lane             2
2778 North Avenue             2
353 Whaley Lane               1
3543 Cherry Ridge Drive       1
3303 Anmoore Road             1
3683 Gorby Lane               1
3781 Hamill Avenue            1
1463 Martha Ellen Drive       1
479 Elmwood Avenue            1
3920 Braxton Street           1
1812 Poplar Street            1
1840 Millbrook Road           1
1009 Felosa Drive             1
4945 Jarvis Street            1
2661 Russell Street           1
1257 Elsie Drive              1
2418 Smith Street             1
720 Tator Patch Road          1
284 Mudlick Road              1
1333 Comfort Court            1
2884 Elsie Drive              1
909 Williams Avenue           1
1094 Jones Avenue             1
1092 Deans Lane               1
2917 Golden Ridge Road        1
4605 Hall Street              1
3209 Crowfield Road           1
2785 Norma Lane               1
                             ..
1092 Far

In [31]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

### Outliers

Check statistics of the dataframes and verify the unexpected outliers

In [9]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [33]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [32]:
treatments[treatments.hba1c_change > 0.90]

given_name      surname auralin    novodra  hba1c_start  hba1c_end  \
1        elliot   richardson       -  40u - 45u         7.56       7.09   
13       gregor         bole       -  47u - 45u         7.61       7.16   
17         gina         cain       -  36u - 36u         7.88       7.40   
20         mile       stanić       -  47u - 48u         7.66       7.24   
25       benoît       bonami       -  44u - 43u         9.82       9.40   
26       suhaim        rahal       -  49u - 47u         7.94       7.50   
27       mizuki        iwata       -  45u - 46u         7.70       7.23   
32        laura  ehrlichmann       -  43u - 40u         7.95       7.46   
40         ásta  grímsdóttir       -  29u - 30u         7.62       7.16   
41       mahmud      kadyrov       -  44u - 43u         7.53       7.11   
49      jackson      addison       -  42u - 42u         7.99       7.51   
53       nasser      mansour       -  33u - 31u         7.51       7.06   
56      juliusz     majewski       -  41u - 39u         7.78       7.31   
57   klementyna   sokołowska       -  42u - 41u         7.98       7.53   
67          joe      edwards       -  36u - 37u         9.35       8.89   
71         hugo      collins       -  53u - 55u         8.00       7.53   
76     noriyuki        sakai       -  32u - 31u         7.58       7.16   
77      placido      udinesi       -  38u - 41u         7.91       7.49   
88       yumena     nakayama       -  34u - 32u         7.76       7.35   
91      michael   kristensen       -  40u - 39u         8.51       8.08   
92      sousuke     nakahara       -  37u - 35u         7.91       7.49   
96    abdul-nur          isa       -  54u - 50u         7.98       7.53   
102         bội           tạ       -  41u - 40u         7.73       7.32   
106       sofia      karlsen       -  35u - 29u         7.62       7.15   
108         mùi        lương       -  52u - 50u         7.74       7.32   
117      javier       moquin       -  30u - 32u         8.00       7.59   
126      jowita   wiśniewska       -  22u - 23u         7.50       7.08   
128       david   gustafsson       -  33u - 34u         7.72       7.28   
133       vasco  van de wiel       -  31u - 29u         7.94       7.51   
138     giovana        rocha       -  23u - 21u         7.87       7.38   
147   krisztián        bakos       -  60u - 60u         7.75       7.29   
153      nicole    zimmerman       -  59u - 56u         8.98       8.57   
154  saidmohmad   rushisvili       -  49u - 46u         7.90       7.48   
155      russom      amanuel       -  39u - 36u         7.86       7.40   
181        kong          lei       -  32u - 30u         7.58       7.15   
182       kajsa        eidem       -  41u - 38u         7.77       7.34   
191      monika       lončar       -  49u - 46u         9.13       8.72   
194   eufrosina    schiavone       -  23u - 23u         9.00       8.56   
202     martina    tománková       -  50u - 57u         7.83       7.36   
213     cecilie       nilsen       -  49u - 50u         9.63       9.21   
215        adib       ghanem       -  46u - 49u         7.59       7.12   
225        liêm         phan       -  43u - 48u         7.58       7.10   
226        daud    batukayev       -  37u - 30u         7.98       7.51   
232      pamela         hill       -  27u - 29u         9.53       9.10   
239       linda        lundy       -  48u - 53u         7.97       7.52   
243     sargent      flamand       -  47u - 39u         7.92       7.49   
245          wu         sung       -  47u - 48u         7.61       7.12   
267        leon       scholz       -  38u - 32u         7.72       7.29   
268      yasmin       araujo       -  51u - 54u         7.82       7.36   
272      steven          roy       -  41u - 43u         7.87       7.43   

     hba1c_change  
1            0.97  
13           0.95  
17           0.98  
20           0.92  
25           0.92  
26           0.94  
27           0.97  
32           0.99  

In [11]:
patients.sample(5)

patient_id assigned_sex given_name    surname                address  \
273         274       female  Mackenzie      McKay  4890 Devils Hill Road   
161         162         male  Mitsunobu   Fujiwara       212 Tibbs Avenue   
502         503         male        Pat    Gersten      2778 North Avenue   
243         244         male    Mustafa     Adonay    1742 Tea Berry Lane   
179         180         male    Dominik  Grunewald    3574 Park Boulevard   

             city     state  zip_code        country  \
273    State Line        MS   39362.0  United States   
161       Chester        MT   59522.0  United States   
502          Burr  Nebraska   68324.0  United States   
243      Superior        WI   54880.0  United States   
179  Marshalltown        IA   50158.0  United States   

                                       contact   birthdate  weight  height  \
273   601-848-3093MackenzieMcKay@superrito.com   3/18/1971   214.9      66   
161  406-759-6160MitsunobuFujiwara@armyspy.com   5/21/1969   223.5      67   
502       PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2      71   
243       715-817-0134MustafaAdonay@dayrep.com    1/5/1922   211.9      67   
179      641-753-5678DominikGrunewald@cuvox.de  11/20/1935   143.0      71   

      bmi  
273  34.7  
161  35.0  
502  19.3  
243  33.2  
179  19.9

In [33]:
#Check patient weights to see any unexpected outliers
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [85]:
#Investigate measure unit issue 
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [86]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

Check that patients have only one of the two treatments. Expect nulls in both querys. 

In [87]:
sum(treatments.auralin.isnull())

0

In [88]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [89]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

##### Define

Add missing records from additional source. 

[documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution

##### Code

In [90]:
df_new = pd.read_csv('treatments_cut.csv')
df_new.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [91]:
treatments_clean = pd.concat([treatments_clean,df_new], ignore_index = True)

##### Test

In [92]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    213 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


##### Define
Fill in the missing HbA1c changes.

##### Code

In [93]:
treatments_clean['hba1c_change'] = treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### Test

In [94]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    350 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


In [95]:
treatments_clean.describe()

hba1c_start   hba1c_end  hba1c_change
count   350.000000  350.000000    350.000000
mean      7.956343    7.560057      0.396286
std       0.545328    0.545456      0.059283
min       7.500000    7.010000      0.200000
25%       7.650000    7.270000      0.350000
50%       7.785000    7.400000      0.390000
75%       7.950000    7.557500      0.440000
max       9.950000    9.580000      0.540000

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email

##### Define

Split the Contact column in patients in order to store the email and phone number in separate columns. 

Regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). 
[regex tutorial](https://regexone.com/). 
[various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). 
[email address regex pattern](http://emailregex.com/)

##### Code

In [96]:
patients_clean.contact.head(10)

0             951-719-9170ZoeWellish@superrito.com
1            PamelaSHill@cuvox.de+1 (217) 569-3204
2                 402-363-6804JaeMDebord@gustr.com
3       PhanBaLiem@jourrapide.com+1 (732) 636-8246
4                  334-515-7487TimNeudorf@cuvox.de
5         386-334-5237RafaelCardosoCosta@gustr.com
6                775-533-5933MaryBAdams@einrot.com
7             XiuxiuChang@einrot.com1 408 778 3236
8          816-265-9578DavidGustafsson@armyspy.com
9    SophieCabreraIbarra@teleworm.us1 718 795 9124
Name: contact, dtype: object

In [97]:
patients_clean['phone_number'] = patients_clean.contact.str.extract(r'((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})',expand=True)

In [98]:
patients_clean['email'] = patients_clean.contact.str.extract(r'([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

In [99]:
patients_clean.drop(['contact'],axis=1,inplace=True)

##### Test

In [100]:
patients_clean.phone_number.sample(5)

482    361-693-4960
299    208-826-1678
459    619-710-6286
359    254-518-6365
260    925-757-6139
Name: phone_number, dtype: object

In [101]:
patients_clean.email.sample(5)

88           MariusHansen@teleworm.us
491    VanessaFerguson@jourrapide.com
150        FelicijanBubanj@einrot.com
397        PipalukPetrussen@rhyta.com
179         DominikGrunewald@cuvox.de
Name: email, dtype: object

In [102]:
# Confirm that no emails start with an integer 
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

##### Define

Normalize the information for the start/end doses and the drug used. 

Pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/)

##### Code

In [103]:
treatments_clean= treatments_clean.replace(['-'], [None])
treatments_clean.head(5)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u       None         7.63       7.20   
1     elliot  richardson       None  40u - 45u         7.56       7.09   
2   yukitaka    takenaka       None  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u       None         7.97       7.62   
4     alissa      montez       None  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

In [104]:
melted = pd.melt(treatments_clean,id_vars=['given_name','surname','hba1c_start','hba1c_end','hba1c_change'],var_name='treatment_type',value_name='dose')
melted = melted.dropna()
melted[melted.surname == 'montez']

given_name surname  hba1c_start  hba1c_end  hba1c_change treatment_type  \
354     alissa  montez         7.78       7.46          0.32        novodra   

          dose  
354  33u - 29u

In [105]:
new_quant = melted.dose.str.split(' - ',expand=True)
melted['start_dose'] = new_quant[0].str[:-1]
melted['end_dose'] = new_quant[1].str[:-1]
melted.head(2)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment_type  \
0   veronika    jindrová         7.63       7.20          0.43        auralin   
3       skye  gormanston         7.97       7.62          0.35        auralin   

        dose start_dose end_dose  
0  41u - 48u         41       48  
3  33u - 36u         33       36

In [ ]:
#alternative to removing the 'u'
treatments_clean.dose_start = treatments_clean.start_dose.str.strip('u')
treatments_clean.dose_start = treatments_clean.end_dose.str.strip('u')

In [106]:
melted.drop(['dose'],axis=1,inplace=True)
melted.head(2)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment_type  \
0   veronika    jindrová         7.63       7.20          0.43        auralin   
3       skye  gormanston         7.97       7.62          0.35        auralin   

  start_dose end_dose  
0         41       48  
3         33       36

In [107]:
treatments_clean = melted
treatments_clean.reset_index(drop=True,inplace=True)

##### Test

In [108]:
treatments_clean.head(5)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment_type  \
0   veronika    jindrová         7.63       7.20          0.43        auralin   
1       skye  gormanston         7.97       7.62          0.35        auralin   
2     sophia      haugen         7.65       7.27          0.38        auralin   
3      eddie      archer         7.89       7.55          0.34        auralin   
4       asia     woźniak         7.76       7.37          0.39        auralin   

  start_dose end_dose  
0         41       48  
1         33       36  
2         37       42  
3         31       38  
4         30       36

##### Define

Add adverse raction information to the treatments table.

[tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution

##### Code

In [109]:
adverse_reactions_clean.head(1)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort

In [110]:
treatments_clean.head(1)

given_name   surname  hba1c_start  hba1c_end  hba1c_change treatment_type  \
0   veronika  jindrová         7.63        7.2          0.43        auralin   

  start_dose end_dose  
0         41       48

In [111]:
treatments_clean = pd.merge(treatments_clean,adverse_reactions_clean, on=['given_name','surname'],how='left')

##### Test

In [112]:
treatments_clean.head(10)

given_name       surname  hba1c_start  hba1c_end  hba1c_change  \
0   veronika      jindrová         7.63       7.20          0.43   
1       skye    gormanston         7.97       7.62          0.35   
2     sophia        haugen         7.65       7.27          0.38   
3      eddie        archer         7.89       7.55          0.34   
4       asia       woźniak         7.76       7.37          0.39   
5     joseph           day         7.70       7.19          0.51   
6    roxanne     andreyeva         9.54       9.14          0.40   
7     simone  baumgaertner         7.74       7.30          0.44   
8       enco        žibrik         7.78       7.34          0.44   
9    camilla      zaitseva         7.53       7.13          0.40   

  treatment_type start_dose end_dose adverse_reaction  
0        auralin         41       48              NaN  
1        auralin         33       36              NaN  
2        auralin         37       42              NaN  
3        auralin         31       38              NaN  
4        auralin         30       36              NaN  
5        auralin         29       36     hypoglycemia  
6        auralin         29       38              NaN  
7        auralin         27       37              NaN  
8        auralin         55       68              NaN  
9        auralin         28       37              NaN

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define

Remove duplicates of the name and surname from multiple tables. The aim is to have the patient id as the key to link the tables. 
Also the name format needs to be uniformized. 

[tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution 

[tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution

##### Code

In [113]:
patients_clean.head(1)

patient_id assigned_sex given_name  surname               address  \
0           1       female        Zoe  Wellish  576 Brown Bear Drive   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   

   height   bmi  phone_number                     email  
0      66  19.6  951-719-9170  ZoeWellish@superrito.com

In [114]:
treatments_clean.given_name = treatments_clean.given_name.str.capitalize()

In [115]:
treatments_clean.surname = treatments_clean.surname.str.capitalize()

In [116]:
treatments_clean.head(5)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment_type  \
0   Veronika    Jindrová         7.63       7.20          0.43        auralin   
1       Skye  Gormanston         7.97       7.62          0.35        auralin   
2     Sophia      Haugen         7.65       7.27          0.38        auralin   
3      Eddie      Archer         7.89       7.55          0.34        auralin   
4       Asia     Woźniak         7.76       7.37          0.39        auralin   

  start_dose end_dose adverse_reaction  
0         41       48              NaN  
1         33       36              NaN  
2         37       42              NaN  
3         31       38              NaN  
4         30       36              NaN

In [117]:
treatments_clean = pd.merge(patients_clean[['patient_id', 'given_name', 'surname']], treatments_clean, on=['given_name','surname'],how='left')

In [118]:
treatments_clean.drop(['given_name','surname'],axis=1,inplace=True)

##### Test

In [119]:
treatments_clean.head(5)

patient_id  hba1c_start  hba1c_end  hba1c_change treatment_type start_dose  \
0           1         7.71        7.3          0.41        novodra         33   
1           2         9.53        9.1          0.43        novodra         27   
2           3          NaN        NaN           NaN            NaN        NaN   
3           4         7.58        7.1          0.48        novodra         43   
4           5          NaN        NaN           NaN            NaN        NaN   

  end_dose adverse_reaction  
0       33              NaN  
1       29              NaN  
2      NaN              NaN  
3       48              NaN  
4      NaN              NaN

In [120]:
#Check duplicate columns
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

15    patient_id
dtype: object

### Quality

##### Define

Change the data type of the zip code to string and add the missing 0 character where needed. 

##### Code

In [220]:
patients_clean.head(2)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   

   height   bmi       phone_number                     email  
0      66  19.6       951-719-9170  ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204      PamelaSHill@cuvox.de

In [135]:
#Convert type
patients_clean.zip_code = patients_clean.zip_code.astype(str)
type(patients_clean.zip_code[0])

str

In [139]:
#Remove the trailing .0
patients_clean.zip_code=patients_clean.zip_code.str.replace('\.0','')

In [221]:
#Fill in 0 for the 4 char ones
patients_clean.zip_code=patients_clean.zip_code.str.pad(width=5,fillchar='0',side='left')

##### Test

In [222]:
patients_clean.zip_code.head(10)

0    92390
1    61812
2    68467
3    07095
4    36303
5    32114
6    84728
7    95037
8    64105
9    10011
Name: zip_code, dtype: object

##### Define

Correct the missing height of one patient (Tim Neudorf)

##### Code

In [231]:
index = patients_clean[patients_clean.height < 30].index
index

Int64Index([4], dtype='int64')

In [232]:
patients_clean.at[index,'height'] = 72

In [ ]:
#Alternative solution
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [233]:
#The query below should not return any results
patients_clean[patients_clean.height < 30]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

##### Define

The state name is not filled in using the same format. Introduce abbreviations for all entries. 

[tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution

##### Code

In [234]:
#check the values we need to change
patients_clean.state.unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'AL', 'Florida', 'NV',
       'CA', 'MO', 'New York', 'MI', 'TN', 'VA', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

In [235]:
#create the mapping dictionary
state_map = {'California': 'CA',
             'New York': 'NY',
             'Illinois': 'IL',
             'Florida': 'FL',
             'Nebraska': 'NE'}

In [237]:
#Create the function to map
def map_state(state):
    if state in state_map.keys():
        abbrev = state_map[state]
        return abbrev
    else:
        return state

In [243]:
patients_clean.state = patients_clean.state.apply(map_state)

##### Test

In [244]:
patients_clean.state.unique()

array(['CA', 'IL', 'NE', 'NJ', 'AL', 'FL', 'NV', 'MO', 'NY', 'MI', 'TN',
       'VA', 'OK', 'GA', 'MT', 'MA', 'NM', 'LA', 'PA', 'CO', 'ME', 'WI',
       'SD', 'MN', 'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD',
       'AZ', 'TX', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH',
       'OR', nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

##### Define

Fix the misspelling of the name of one patient (David Gustafsson).


##### Code

In [245]:
patients_clean[patients_clean.given_name == 'Dsvid']

patient_id assigned_sex given_name     surname             address  \
8           9         male      Dsvid  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

In [264]:
patients_clean.given_name=patients_clean.given_name.replace('Dsvid','David')

##### Test

In [265]:
patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define

Change the data type to categorical where needed. Change the start/end doses to integers. 

[documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution

[documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution

[documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution

##### Code

In [266]:
#create the categorical variables
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

In [267]:
#Convert column to date time 
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

In [276]:
#remove NANs from the dose columns
treatments_clean.start_dose= treatments_clean.start_dose.fillna(0)
treatments_clean.end_dose= treatments_clean.end_dose.fillna(0)

In [277]:
#Convert columns to int
treatments_clean.start_dose = treatments_clean.start_dose.astype(int)
treatments_clean.end_dose = treatments_clean.end_dose.astype(int)

##### Test

In [278]:
patients_clean.dtypes

patient_id               int64
assigned_sex          category
given_name              object
surname                 object
address                 object
city                    object
state                 category
zip_code                object
country                 object
birthdate       datetime64[ns]
weight                 float64
height                   int64
bmi                    float64
phone_number            object
email                   object
dtype: object

In [279]:
treatments_clean.dtypes

patient_id            int64
hba1c_start         float64
hba1c_end           float64
hba1c_change        float64
treatment_type       object
start_dose            int32
end_dose              int32
adverse_reaction     object
dtype: object

##### Define

Change number format to have uniform representation for all entries in the dataframe. 

[Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [280]:
#Remove all special characters from the phone number
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '')

In [281]:
#Fill in country code where missing
patients_clean.phone_number = patients_clean.phone_number.str.pad(11,fillchar='1')

##### Test

In [282]:
patients_clean.phone_number.head(5)

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

##### Define

Remove John Doe data since it is assumed this is not recoverable. 

##### Code

In [283]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [284]:
patients_clean[patients_clean.surname=='Doe']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

##### Define

Remove the duplicate entries for 3 of the patients: Jakobsen, Gersten, Taylor.

##### Code

In [290]:
patients_clean = patients_clean[~(patients_clean.address.duplicated()& patients_clean.address.notnull())]

##### Test

In [294]:
patients_clean[(patients_clean.surname == 'Jakobsen')|(patients_clean.surname == 'Gersten')|(patients_clean.surname == 'Taylor')  ]

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
97           98         male    Patrick   Gersten   2778 North Avenue   
131         132       female     Sandra    Taylor  2476 Fulton Street   
426         427         male    Rogelio    Taylor  4064 Marigold Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
97          Burr    NE    68324  United States 1954-05-03   138.2      71   
131     Rainelle    WV    25962  United States 1960-10-23   206.1      64   
426        Miami    FL    33179  United States 1992-09-02   186.6      69   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
97   19.3  14028484923      PatrickGersten@rhyta.com  
131  35.4  13044382648      SandraCTaylor@dayrep.com  
426  27.6  13054346299    RogelioJTaylor@teleworm.us  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

##### Define

Correct the measure unit for one patient: Zaitseva which has her weight in kgs instead of lbs

##### Code

In [298]:
patients_clean[patients_clean.weight==patients_clean.weight.min()]

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate  weight  height   bmi  \
210  Wooster    OH    44691  United States 1938-11-26    48.8      63  19.1   

    phone_number                          email  
210  13302022145  CamillaZaitseva@superrito.com

In [303]:
weight = patients_clean.weight.min()
surname = patients_clean[patients_clean.weight==patients_clean.weight.min()]['surname']

In [305]:
mask = patients_clean.surname == 'Zaitseva'

In [307]:
patients_clean.loc[mask,'weight'] = weight * 2.20462

##### Test

In [308]:
patients_clean[patients_clean.weight==patients_clean.weight.min()]

patient_id assigned_sex given_name surname           address       city  \
459         460       female     Idalia   Moore  4380 Grim Avenue  San Diego   

    state zip_code        country  birthdate  weight  height   bmi  \
459    CA    92073  United States 1993-10-26   102.1      61  19.3   

    phone_number                  email  
459  16197106286  IdaliaEMoore@cuvox.de